### PR TITLE
DiscoveryNode and TransportAddress should preserve host information

### PR DESCRIPTION
**Pull request by jaymode**
In some cases, such as the creation of DiscoveryNode instances for unicast ping requests, the host information was not being populated properly and instead the address string was being used. Additionally, when serializing a DiscoveryNode and in turn a transport address, the host was not being set on the InetAddress when deserializing the object, so even if the address was created from a hostname, the address in the deserialized instance had no knowledge of the hostname that was originally used.

### CODE REVIEW

File: core/src/test/java/org/elasticsearch/cluster/node/DiscoveryNodeTests.java [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+
+import static java.util.Collections.emptyMap;
+import static java.util.Collections.emptySet;
+
+public class DiscoveryNodeTests extends ESTestCase {
+
+    public void testDiscoveryNodeIsCreatedWithHostFromInetAddress() throws Exception {
+        InetAddress inetAddress = randomBoolean() ? InetAddress.getByName("192.0.2.1") :
+            InetAddress.getByAddress("name1", new byte[] { (byte) 192, (byte) 168, (byte) 0, (byte) 1});
+        TransportAddress transportAddress = new TransportAddress(inetAddress, randomIntBetween(49152, 65535));
```
Comment (jasontedor) [inline core/src/test/java/org/elasticsearch/cluster/node/DiscoveryNodeTests.java]: I'm just curious why you use a random ephemeral port here (and in the other tests)?

**Review by jasontedor — COMMENTED** [review]
I left a question.

File: core/src/test/java/org/elasticsearch/cluster/node/DiscoveryNodeTests.java (hunk 1 above)
Comment (jasontedor) [inline core/src/test/java/org/elasticsearch/cluster/node/DiscoveryNodeTests.java]: Yeah, let's do that (expand the range to the entire valid port range, and add assertions etc.). Thanks @jaymode.

**Review by jasontedor — APPROVED** [review]
LGTM.

**Comment by s1monw** [comment]
LGTM 2

**Pull request merged**